### DISCUSSION TITLE
How to define rust type of TIMESTAMP?

### DISCUSSION THREAD

**Discussion by lz1998** [SeaQL/sea-orm#341] [Q&A] [answered]
I have tried 4 types, never of them works. **`NaiveDateTime`, `i64`, `DateTime<Utc>` and `DateTime<FixedOffset>`**

I use MySQL 8. I create table by myself without `automigrate`, and I define rust struct by myself.

```SQL
    `created_at` TIMESTAMP    NOT NULL DEFAULT CURRENT_TIMESTAMP,
    `updated_at` TIMESTAMP    NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP
```

```rust
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "product")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: i64,
    pub name: String,
    pub operator: String,
    pub status: i32,
    pub created_at: DateTime<FixedOffset>,
    pub updated_at: DateTime<FixedOffset>,
}

#[derive(Copy, Clone, Debug, EnumIter)]
pub enum Relation {}

impl RelationTrait for Relation {
    fn def(&self) -> RelationDef {
        panic!("No RelationDef")
    }
}

impl ActiveModelBehavior for ActiveModel {}
```

**`NaiveDateTime` and `i64`**
```bash
Err(Query("error occurred while decoding column \"created_at\": mismatched types; Rust type `core::option::Option<chrono::naive::datetime::NaiveDateTime>` (as SQL type `DATETIME`) is not compatible with SQL type `TIMESTAMP`"))
```

**`DateTime<Utc>` and `DateTime<FixedOffset>`**
```bash
error: proc-macro derive panicked
 --> src/database/product.rs:8:35
  |
8 | #[derive(Clone, Debug, PartialEq, DeriveEntityModel, Serialize, Deserialize)]
  |                                   ^^^^^^^^^^^^^^^^^
  |
  = help: message: `"DateTime<Utc>"` is not a valid identifier

```

**Comment by billy1624** [comment]
Hi @lz1998, welcome!!

Can you try `DateTime` or `DateTimeWithTimeZone`? It was re-exported here.
https://github.com/SeaQL/sea-orm/blob/cc4904993a190e827f3031210bd965e6f46c9165/src/entity/prelude.rs#L24-L29

**Reply by lz1998** [reply]
Should I add some features in Cargo.toml?
```toml
sea-orm = { version = "0.4.0", features = ["sqlx-mysql", "runtime-tokio-native-tls", "macros", "with-chrono"], default-features = false }
```


**DateTime**
```bash
Err(Query("error occurred while decoding column \"created_at\": mismatched types; Rust type `core::option::Option<chrono::naive::datetime::NaiveDateTime>` (as SQL type `DATETIME`) is not compatible with SQL type `TIMESTAMP`"))
```

**DateTimeWithTimeZone**
```bash
chrono::DateTime<chrono::FixedOffset> unsupported by sqlx-mysql
```

**Reply by billy1624** [reply]
As you can see here, `chrono::DateTime<chrono::FixedOffset>` is not supported by SQLx MySQL
- https://docs.rs/sqlx/latest/sqlx/struct.MySql.html

I guess we need to support one of ... in sea-orm to make this work
- `chrono::DateTime<chrono::Utc>`
- `chrono::DateTime<chrono::Local>`

I will try to fix this later today. Stay tuned!

**Reply by lz1998** [reply]
Is there a branch of nightly version?

**Reply by billy1624** [reply]
You mean `sea-orm`? I will create a PR fixing this, then you can update the toml as something like this
```toml
sea-orm = { version = "0.4.0", git = "https://github.com/SeaQL/sea-orm.git", branch = "some-branch-name", ... }
```
Does this make sense to you?

**Reply by lz1998** [reply]
If I want to get a i64 timestamp of seconds, what should I do?

**Reply by billy1624** [reply]
You need to cast the value of timestamp column to integer inside the SQL statement.

**Comment by billy1624** [comment, marked as answer]
For anyone who was watching this discussion. A issue was created for this discussion.
- #344
